src/computability/language.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson.
+-/
+import data.finset.basic
+
+/-!
+# Languages
+This file contains the definition and operations on formal languages over an alphabet. Note strings
+are implemented as lists over the alphabet.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- A language is a set of strings over an alphabet. -/
+@[derive [has_union, has_emptyc, has_mem (list α), has_singleton (list α)]]
```

Comment:
```suggestion
@[derive [has_mem (list α), has_singleton (list α), has_insert (list α)]]
```
As I've written below, I think we should drop these in favor of `0` and `+` that we already have from the semiring instance.

The `has_insert` instance makes it possible to write `({[tt], [tt,ff]} : language bool)`, which is neat.

Reply:
Is it possible to derive the instances in just this file, or at least a better way of writing things like `instance : inhabited (language α) := ⟨begin rw language, exact ∅ end⟩`

Reply:
I'd go for `(∅ : set _)`.

Reply:
It might be enough to add `local attribute [reducible] language` in the section where you define `zero` `add` etc. For `inhabited` you should probably define it in terms of `0` anyway.

Reply:
I'm having trouble with addition now, I have defined `instance : has_add (language α) := ⟨set.union⟩` and `lemma add_def (l m : language α) : l + m = set.union l m := rfl` but it seems mathlib has very limited support for `set.union` and I can't figure out how to limit the uses of `∪`.

Reply:
Definitely avoid `set.union`.  Either use `(l ∪ m : set _)`, or instead keep `(∪)` but add `l + m = l ∪ m` as a simp lemma.

Reply:
This is a type synonym, so you should minimize the use of functions that see through that. Just use the has_mem instance or make a coercion to set.